### ISSUE TITLE
Plugin auto discovery & plugin registry

### ISSUE THREAD

**Issue by McFoggy** [gitbucket/gitbucket#1468]
This issue is created to expose/gather needs/idea for a future implementation of a plugin discovery/installation system in gitbucket probably working together with a plugin registry.

# Requirements to provide a plugin installation mechanism in gitbucket

## with internet access

- a central registry exists
- the registry exposes API to retrieve plugin information
    + plugin name/description/version/download URL
- with hotdeployment
    + gitbucket download the new plugin/version
    + gitbucket/plugins stop/unload old plugin classes & services
    + gitbucket deploy the downloaded plugin into GITBUCKET_HOME/plugin-name.jar erasing the previously installed one (if one existed)
    + gitbucket creates an URLClassLoader for this plugin and initialize it
- without hotdeployment
    + gitbucket download the new plugin/version
    + gitbucket stores it in a temporary place
    + on next reload, before starting plugins, gitbucket deploy the plugin from temporary place into GITBUCKET_HOME/plugin-name.jar erasing the previously installed one (if one existed)
    + gitbucket creates an URLClassLoader for this plugin and initialize it

## without internet access

- gitbucket server does not see the internet, but user does
- a central registry exists
- the registry exposes page to select plugins/version to installation
- a download action is available to collect a packaging of the selected plugins
- user put the downloaded package in some location
- on startup or with hotdeployment gitbucket is capable of extracting plugins from this downloaded package

### Pure static registry

Can be a simple website exposing static data (json, xml, yaml) via for example github pages.

__PRO__: simple, fast to setup, good enough to start, running cost is 0  
__CONS__: need pure manual update, cannot gather/expose feedback from gitbucket installations

### Static registry updated automatically

Same as the pure static registry with some automation to update the data.
Could be some implemented with some build script running continuous integration systems (like travis-ci, circleci, ...) to execute some logic to update the static data
- for each registered plugin
    - fetch its latest data (via github API, ...)
    - update the static data file
   - commit/push them back

__PRO__: provide dynamic behavior, running cost is 0  
__CONS__: need a CI service offering "cron" like feature, cannot gather/expose feedback from gitbucket installations

### Dynamic registry 

The registry is a webapplication hosted somewhere.

__PRO__: cover every possible scenario including dynamic feedback from gitbucket installations (number of installations, plugin incompatibilities, starring, ...)
__CONS__: requires CPU Power + data storage

**Comment by takezoe** [comment]
We might possible to use the maven central as the plugin jar repository. We only need to provide a plugin list on an other places (e.g. gitbucket community plugins) statically.

**Comment by aadrian** [comment]
> We might possible to use the maven central as the plugin jar repository.

+ 1. 

> ... cannot gather/expose feedback from gitbucket installations

Sonatype also has download statistics, so it's possible to know what plug-ins are the most used.

**Comment by McFoggy** [comment]
To clarify for me the registry was not a repository, ie the jars would not be there, but only a central storage/provider of metadatas about the plugins like: id, name, versions, hosting type for binaries (maven, github, ...).
I would let the storage story completly out of gitbucket ; even the computation of downloadable links would be invisible for gitbucket, only the registry would know, depending on the plugin hosting type to compute a URL for a particular version.

**Comment by takezoe** [comment]
@McFoggy Basically agree. How do you think about containing downloadable url for each versions of plugin to metadatas instead of  the computation (from hosting type and version?) as an easier and more flexible way?

**Comment by McFoggy** [comment]
@takezoe in the pure static registry scenario, basically we would just expose some json/xml files containing already all infos including downloadable URLs.
Having the capability to compute things (dynamically or by cron) allows the system to _discover_ new versions ; but that could be  _just_ an enhancement that comes later on.

**Comment by McFoggy** [comment]
I have created quickly a possible API specification that could be afterwards extended.

Visit the following:
- https://anypoint.mulesoft.com/apiplatform/n-a-307/#/portals/organizations/412d179f-5107-4b04-a8ab-a76280294391
- click on version 1 of Gitbucket registry API
- on the left click on API reference
- navigate inside the API definition

@takezoe I have invited you to collaborate on this mulesoft API project so that you can also edit the RAML specification.

**Comment by McFoggy** [comment]
I just pushed in gitbucket-plugins some demo files so that some URL are queryable
- http://gitbucket-plugins.github.io/registry/plugins.json
- http://gitbucket-plugins.github.io/registry/plugins.yaml
- http://gitbucket-plugins.github.io/registry/plugins/emoji/versions
- http://gitbucket-plugins.github.io/registry/plugins/emoji/versions.json
- http://gitbucket-plugins.github.io/registry/plugins/emoji/versions.yaml

drawback is that using this technic:
- we cannot host inside registry a file called `plugins` and a directory called `plugins`
- this would imply that either
    - we use file extension, not very REST compatible
    - or we change a bit the API
        - `registry/plugins` with a plural as a file
        - `registry/plugin` singular as a directory to be able to serve sub domains

**Comment by aadrian** [comment]
@McFoggy 
Please also add at least:
  - Plug-in repo url - e.g. the GitHub project URL
  - GitBucket Version Range compatibility
  - GitBucket Version it was compiled against

Also since the JARs would be hosted inside a maven, repo, IMO the versions.json should not contain the Download URL, but only the Maven Identification params.

**Comment by McFoggy** [comment]
@aadrian 
> Also since the JARs would be hosted inside a maven, repo, IMO the versions.json should not contain the Download URL, but only the Maven Identification params.

Sorry but here I do not agree, especially if we think in long term with a registry that will gain some logic at some point in time. 
For me the API we are discussing here is the API that is going to be consumed by the gitbucket app. It should contain "resolved" data and we should not, in my opinion, let too much computations occur on the API client side.
In the future, I hope we will be able to declare in the registry that plugin X uses Maven as file repository, that plugin Y uses GitHub releases and still expose in the API already computed/resolved URL for those plugins download.

For the version compatibility, I agree there should be something in the global picture that handles the topic. 
We can give your proposal a trial , but I am sure that we will have in the end to enhance gitbucket in a way it exposes a Plugin API version with compatibility scheme. 
It should be against this Plugin API version that plugins should express their compatibility not against a gitbucket version. Moreover, we should not have to ask plugin maintainers to perform checks/update/actions when a new gitbucket version is out. In the end we should be able to compute those compatibilities/incompatibilities automatically.

**Comment by McFoggy** [comment]
@aadrian 
I have updated the model & samples to include your remarks on project URL, compile version & version range.

**Comment by aadrian** [comment]
> Sorry but here I do not agree, especially if we think in long term with a registry that will gain some logic at some point in time.

@McFoggy a hardcoded download path means that one needs DNS access or some sort of load balancing on the long run.

Why not let Maven/SBT/Gradle make the download? They use only those IDs and a list of servers.

> It should be against this Plugin API version that plugins should express their compatibility not against a gitbucket version. 

Sure, but that is a long shot and would require quite allot of discipline :) .

**Comment by McFoggy** [comment]
> Why not let Maven/SBT/Gradle make the download? They use only those IDs and a list of servers.

Not sure we are talking about the same thing. We are not going to embed maven, sbt, gradle or ivy inside gitbucket runnable.
The registry we are talking about will  be consumed by a gitbucket installation at runtime ; we are not talking about any build tools.

For the beginning, I do not see the exposure of direct links to github releases or maven central as a problem (for the scenario with internet access). Here we are just talking about downloading plugins from a gitbucket admin screen ; it is an action that will be done only from time to time by our users. I would not complicate too much now this URL stuff ; premature optimization is not required here (in my opinion).

**Comment by aadrian** [comment]
> We are not going to embed maven, sbt, gradle or ivy inside gitbucket runnable.

Well, the part that "sloves" paths and dependencies and does the download, considering proxies too, should be quite reusable. 

For the "offline" scenario (IMHO the most used case for GitBucket), the "client" part (i.e. what runs on the users computer with Internet connection could be simply a Gradle or SBT script or something based on it.

**Comment by McFoggy** [comment]
For the pure static or semi-static generation, I didn't thought of it initially but we can leverage github pages & jekyll from gitbucket-plugins site to generate some fancy stuff.

**Comment by takezoe** [comment]
@McFoggy Thanks for your work. Demo files look good, but I have some suggestion:

- plugins
  - Add author (or provider) and its url (e.g. GitHub account page or corporate web site)
- versions
  - gitbucketCompatibleRange looks redundant. In particular, is include is needed?

**Comment by McFoggy** [comment]
@takezoe 

> gitbucketCompatibleRange looks redundant
Yes and no. It depends where we want to put the logic. As I said previously in this post I think there should be in gitbucket world a concept of Plugin API version compatibility. Will it be handled by a real plugin API version, by a serie of gitbucket version, I do not know but the concept should be there somehow.
Then regarding this `gitbucketCompatibleRange` field its existence depends on where we want to put the logic:
- is it to gitbucket app to know/calculate the range of gitbucket(or plugin API) compatible versions?
-  or is it to the registry to handle this logic and thus to expose it as a result on its API?

If it is the former, then the registry (and thus plugins version) should only expose the Plugin API version there are compatible with or the gitbucket version they were build against.

**Comment by takezoe** [comment]
@McFoggy I meant `gitbucketCompatibleRange` is required, but `include` attribute might be unnecessary. I think there is no problem by handling the declared version as `include` always.

**Comment by McFoggy** [comment]
A little status on this topic.

I have quickly implemented a generation mechanism (using [nodejs](https://github.com/gitbucket-plugins/gitbucket-plugins.github.io/blob/master/package.json) & [travis-ci](https://github.com/gitbucket-plugins/gitbucket-plugins.github.io/blob/master/.travis.yml)) which parses a [plugins.yaml](https://github.com/gitbucket-plugins/gitbucket-plugins.github.io/blob/master/plugins.yaml) data file to generate the files as specified in the [API spec](https://anypoint.mulesoft.com/apiplatform/n-a-307/#/portals/organizations/412d179f-5107-4b04-a8ab-a76280294391/apis/11605863/versions/156366/pages/206694). _Notice that I have changed the generated URL endpoint from `registry/plugins/pluginID`  to `registry/plugin/pluginID` (before plural, now singular) to avoid naming collision on the generated file system._

Now we are able to update the registry files by "just" updating the yaml file and pushing it back to git.

Then once [travis-ci](https://travis-ci.org/gitbucket-plugins/gitbucket-plugins.github.io) has done his work we can query the registry:

`wget -qO- http://gitbucket-plugins.github.io/registry/plugins | head -n 20`
```
[
  {
    "id": "h2-backup",
    "name": "H2 Database Backup Plugin",
    "description": "Allows to export h2 database of gitbucket",
    "projectURL": "https://github.com/gitbucket-plugins/gitbucket-h2-backup-plugin",
    "versions": [
      {
        "version": "1.4.0",
        "downloadURL": "https://github.com/gitbucket-plugins/gitbucket-h2-backup-plugin/releases/download/1.4.0/gitbucket-h2-backup-plugin.jar",
        "gitbucketCompileVersion": "4.10.0",
        "gitbucketCompatibleRange": {
          "lower": {
            "bound": "4.10.0",
            "included": true
          }
        }
      }
    ]
  },
```

same applies for all exposed plugins

` wget -qO- http://gitbucket-plugins.github.io/registry/plugin/emoji/versions`

```
[
  {
    "version": "4.4.0",
    "downloadURL": "https://github.com/gitbucket/gitbucket-emoji-plugin/releases/download/4.4.0/gitbucket-emoji-plugin_2.12-4.4.0.jar",
    "gitbucketCompileVersion": "4.10.0",
    "gitbucketCompatibleRange": {
      "lower": {
        "bound": "4.10.0",
        "included": true
      }
    }
  }
]
```

We can thus start consuming it.

**Comment by aadrian** [comment]
> We can thus start consuming it.

This looks great!

Thank you very much :) .

**Comment by McFoggy** [comment]
from comments in #1484, I'd suggest that we start first with the scenario _with internet access/without hot deployment_. 
Of course this will enforce a restart of the server like today, but it will already ease installation & awareness of new plugins.

**Comment by takezoe** [comment]
@McFoggy In my opinion, Hot deployment is more important feature because it will be the base of plugin development and installation. I started a hot deployment in #1487, but it takes a while.

We can progress with the scenario _with internet access/without hot deployment_ apart from that because this scenario is responsible for plugin discovering and downloading. I think we can merge them later.

**Comment by aadrian** [comment]
@McFoggy @takezoe 
Would "plug-in dependencies" (i.e. what **other** plug-ins does this plug-in require or depends on) also be in the scope ?

**Comment by McFoggy** [comment]
@aadrian of course it would be cool to have out of the box plugin dependencies but as a first step perhaps we can consider that plugin come with all dependencies packaged (like today). Then later on, we'll have to enhance our system to support transitive dependencies.  
But the later means also that we will need to either reference/store those dependencies already calculated (either in the registry or in plugins artifacts) or that we will have to use plugin project build files (build.sbt, pom.xml, build.gradle, ivy.xml, ...) to resolve the dependencies.  
Enhancements could also have to be done also on the classloading system in order to avoid collisions with gitbucket core dependencies.
Finally I would suggest to create another issue to track this feature.

**Comment by aadrian** [comment]
@McFoggy 
> but as a first step perhaps we can consider that plugin come with all dependencies packaged (like today).

My **Use Case** would be more when a plug-in ```plugin_B:v1.0``` depends on another plug-in ```plugin_A:v2.0```. 

E.g. a **new** plug-in that would need  the PlantUML plug-in installed and working in order for itself be able to work correctly.

At the moment this can't be specified, and it' the sole responsibility of the user to take care of everything (and this is quite error prone).

If this new YAML plug-in descriptor could also have a **```depends```** section, than at least the user could be warned by the UI that it needs to download other plug-ins first, or if it disables a plug-in, other would be affected too.

----

> Then later on, we'll have to enhance our system to support transitive dependencies.

Regarding ```jar``` dependencies, that would be indeed something more complicated, but could allow GitBucket itself to be a "collection of plug-ins" :). 

However, there are projects that seem to solve this already: http://www.eclipse.org/aether/ and even SBT seems to use it now too: https://github.com/sbt/sbt/pull/1793  :) .

**Comment by McFoggy** [comment]
@aadrian makes sense for plugin dependencies

regarding jars dependencies _(being one of the creator of Ivy at Jayasoft before we donated it to Apache foundation, I know the subject can be tricky)_ if we could stick to fat jars with classloader isolation ; I think it will ease our life for the moment.

**Comment by kounoike** [comment]
I wrote small [test](https://github.com/kounoike/gitbucket-ipynb-plugin-test).
It checks the target plugin really loaded to GitBucket. And it can run by [travis-ci](https://travis-ci.org/kounoike/gitbucket-ipynb-plugin-test/builds/219957142).

I think it is useful for check which plugin can be used with new-version of GitBucket with prebuild binary.
If check is passed, automatically upload plugin-registry info to server by some script on travis.

**Comment by kounoike** [comment]
@McFoggy I think plugin-ordered list is not useful for GitBucket plugin discovery system.
It represents only "what versions are available for specified plugin".

In GitBucket, it is required "what plugins are available for this GitBucket version".

I think it is better for `https://some.server/some/where/registry/for_4.11.0.json`:
```
[
  {
    "name": "emoji",
    "version": "x.xx",
    "url": "https://..."
  },
 ...
]
```

or
```
{
  "emoji": { "version": "x.xx", "url": "https://..." },
  "gist": ...
}
```

I think it is easily generatable by script with travis-ci compatibility check(commented above), and easy to use by GitBucket.

**Comment by kounoike** [comment]
I make two repositories for plugin-registry. (Maybe it can be merge to one repo.)

- [gitbucket-plugin-checker](https://github.com/kounoike/gitbucket-plugin-checker)
- [gitbucket-plugin-registry](https://github.com/kounoike/gitbucket-plugin-registry)

these repositories with travis-ci act as:

1. gitbucket-plugin-checker checks each plugin can be loaded to specified version of GitBucket.
2. if check passed, travis-ci automatically pushes individual plugin/GBversion information to gitbucket-plugin-registry repositry's master branch. ex. [emoji.json](https://github.com/kounoike/gitbucket-plugin-registry/blob/master/registry/plugins/4.11.0/emoji.json)
3. gitbucket-plugin-registry's travis aggregates all plugin's information to single json file and push to `registry` branch. ex. [4.11.0.json](https://github.com/kounoike/gitbucket-plugin-registry/blob/registry/release/4.11.0.json)
4. finally, we can use 4.11.0.json for GitBucket 4.11.0's plugin registry file.

*for GitBucket release*:
- change [gitbucket-version](https://github.com/kounoike/gitbucket-plugin-checker/blob/master/gitbucket-version) to new version, new URL.
- commit and push it.
- automatically travis-ci checks all plugin's compatibility and generates new version's plugin registry file.

*for Plugin authors (if check is failed)*:
- update plugin with new GitBucket version and release it.
- create PR with change plugin's description file ex. [plugins/emoji](https://github.com/kounoike/gitbucket-plugin-checker/blob/master/plugins/emoji).
- When Plugin-registry maintainer merge the PR, new registry file is generated automatically.

How about this kind of mechanism.

**Comment by takezoe** [comment]
Plugin hot deployment (#1487) has been merged. It supports local plugin repository for bundled plugins. It read plugins information from [plugins.json](https://github.com/gitbucket/gitbucket/blob/master/src/main/resources/plugins/plugins.json). I would like to add support for remote repository as same (or maybe extended) way.

**Comment by kounoike** [comment]
[This conversation(Japanese)](https://twitter.com/takezoen/status/969370077190529024) says:

- We prefer build **all** plugins for new GitBucket version
- We prefer **free** hosting service

so, I wrote travis-based plugin building farm.

https://github.com/kounoike/gitbucket-plugin-farm-test

- It depends on sbt-gitbucket-plugin
- Build all plugin for new GitBucket version **on this farm**
- It automatically generate `plugins.json` for new GitBucket version.
- Publish plugin jar and `plugins.json` to GitHub release

Please tell me if you have any opinion.
